Modules/arraymodule.c

```diff
@@ -2976,42 +2964,48 @@ static PyMethodDef a_methods[] = {
 static int
 array_modexec(PyObject *m)
 {
+    array_state *state = get_array_state();
     char buffer[Py_ARRAY_LENGTH(descriptors)], *p;
     PyObject *typecodes;
     const struct arraydescr *descr;
 
-    if (PyType_Ready(&Arraytype) < 0)
+    state->ArrayType = (PyTypeObject *)PyType_FromModuleAndSpec(m, &array_spec, NULL);
```

Comment:
Since the array extension uses the multi-phase initialization API, array_modexec() can be called multiple times. Problem: with your change, each time the state is overriden. The minimum fix would be to not override state if it's already initialized.

The best would be to have a state per module, but then the problem become how to retrieve the module state from array type methods. For methods declared with Argument Clinic, the new "defining_class" type can be used. But for some functions like array_concat() (Py_sq_concat slot), you only have access to an instance. Right now, it's not possible to safely retrieve the module state from a type. Here, array.array is a "base type" (Py_TPFLAGS_BASETYPE) and so can be subclassed. If it wouldn't be a base class, it would be safe to simply use PyType_GetState().

Here you need to wait until @encukou implements his function traversing the MRO to retrieve the parent type with the expected state => https://bugs.python.org/issue42100

cc @corona10 @shihai1991 

I'm not sure that your change enhances the code or makes it worse :-(

Reply:
> Since the array extension uses the multi-phase initialization API, array_modexec() can be called multiple times. Problem: with your change, each time the state is overriden. The minimum fix would be to not override state if it's already initialized.

Ah, of course. Yes, that's an easy temporary workaround.

> The best would be to have a state per module, but then the problem become how to retrieve the module state from array type methods.

Yes, I've encountered this while working on bpo-42064. I've temporarily worked around it by using basic MRO traversal:
```
PyTypeObject *_pysqlite_find_base_type(PyObject *self, const char *type_name) {
      PyTypeObject *self_tp = Py_TYPE(self);
      PyObject *bases = (self_tp != NULL) ? self_tp->tp_bases : NULL;
      PyTypeObject *tp = NULL;
      int i;
  
      for (i = 0; i < PyTuple_Size(bases); i++) {
          PyTypeObject *t = (PyTypeObject *)PyTuple_GetItem(bases, i);
  
          if (t != &PyBaseObject_Type && strcmp(t->tp_name, type_name) == 0) {
              tp = t;
              break;
          }
      }
  
      if (tp == NULL) {
          tp = self_tp;
      }
  
      return tp;
}
```

> Here you need to wait until @encukou implements his function traversing the MRO to retrieve the parent type with the expected state => https://bugs.python.org/issue42100

I noticed his work, but I haven't had the chance to take a look at it yet. It would be great news if a proper MRO traversal mechanism was available!

> I'm not sure that your change enhances the code or makes it worse :-(

Me neither :) Nevertheless, the effort is wort it for my part (learning the C API, getting to know different parts of the code base, etc.)

Thanks for helpful and quick feedback! 🙏🏻

Reply:
Would something like 807fcb5 be an acceptable workaround, until the module state issues are resolved?

Reply:
Now that 57aaaa8d2a43bddffeca4ead5632ce6652945cc0 got in, I'll rebase onto master and use  `_PyType_GetModuleByDef` to establish a module state. I guess this should go to bpo-1635741 instead, then?

Reply:
FYI, @vstinner: rebased onto master and applied `defining_class` and `_PyType_GetModuleByDef` to help fetch module state.